inc/class-wpseo-utils.php

```diff
@@ -483,35 +495,276 @@ public static function clear_transient_cache( $option ) {
 		$wpdb->query( $query );
 	}
 
 	/**
+	 * Get the sitemap cache key prefix
+	 *
+	 * @return string
+	 */
+	public static function get_sitemap_cache_key_prefix() {
+		return self::$sitemap_cache_key_prefix;
+	}
+
+	/**
+	 * Get the cache key for a certain type and page
+	 *
+	 * Example key format for type "post", page 1: wpseo_sitemap_post_1:akfw3e_23azBa
+	 *
+	 * @param null|string $type The type to get the key for. Null or self::SITEMAP_INDEX_TYPE for index cache.
+	 * @param int         $page The page of cache to get the key for.
+	 *
+	 * @return string The key where the cache is stored on.
+	 */
+	public static function get_sitemap_cache_key( $type = null, $page = 1 ) {
+		// Using SITEMAP_INDEX_TYPE for sitemap index cache.
+		$type = is_null( $type ) ? self::SITEMAP_INDEX_TYPE : $type;
+
+		$global_cache_validator = self::get_sitemap_cache_validator();
+		$type_cache_validator   = self::get_sitemap_cache_validator( $type );
+
+		$prefix = self::$sitemap_cache_key_prefix;
+		$postfix = sprintf( '_%d:%s_%s', $page, $global_cache_validator, $type_cache_validator );
+
+		$type = self::get_safe_sitemap_cache_type( $type, $prefix, $postfix );
+
+		// Build key.
+		$full_key = $prefix . $type . $postfix;
+
+		return $full_key;
+	}
+
+	/**
+	 * If the type is over length make sure we compact it so we don't have any database problems
+	 *
+	 * When there are more 'extremely long' post types, changes are they have variations in either the start or ending.
+	 * Because of this, we cut out the excess in the middle which should result in less chance of collision.
+	 *
+	 * @param string $type The type of sitemap to be used.
+	 * @param string $prefix The part before the type in the cache key. Only the length is used.
+	 * @param string $postfix The part after the type in the cache key. Only the length is used.
+	 *
+	 * @return string The type with a safe length to use
+	 *
+	 * @throws OutOfRangeException When there is less than 15 characters of space for a key that is originally longer.
+	 */
+	private static function get_safe_sitemap_cache_type( $type, $prefix = '', $postfix = '' ) {
+		// Length of key should not be over 53.
+		$max_length = 53;
+		$max_length -= strlen( 'timeout_' );
+		$max_length -= strlen( $prefix );
+		$max_length -= strlen( $postfix );
+
+		if ( strlen( $type ) > $max_length ) {
+
+			if ( $max_length < 15 ) {
+				/**
+				 * If this happens the most likely cause is a 'page' that is too big.
+				 * "Normally" the max_length is 24 + strlen( page ), which is unlikely to go above 10 in the first place.
+				 *
+				 * So this would not happen unintentionally..
+				 * Either by trying to cause a high server load, finding backdoors or misconfiguration.
+				 */
+				throw new OutOfRangeException( 'Trying to build a safe sitemap cache key, but the postfix and prefix combination leaves too little room to do this. You are probably requesting a page that is way out of expected range.' );
+			}
+
+			$half = ( $max_length / 2 );
```

Comment:
I'm unsure what you're trying to achieve here (lines 653-658) could you please elaborate?